src/algorithms/convex.rs

```diff
@@ -0,0 +1,293 @@
+//! Convexity checking for portgraphs.
+//!
+//! This is based on a [`ConvexChecker`] object that is expensive to create
+//! (linear in the size of the graph), but can be reused to check multiple
+//! subgraphs for convexity quickly.
+
+use std::collections::BTreeSet;
+
+use bitvec::bitvec;
+use bitvec::vec::BitVec;
+
+use crate::algorithms::toposort;
+use crate::{Direction, LinkView, NodeIndex, PortIndex, SecondaryMap, UnmanagedDenseMap};
+
+use super::TopoSort;
+
+/// A pre-computed datastructure for fast convexity checking.
+pub struct ConvexChecker<G> {
+    graph: G,
+    // The nodes in topological order
+    topsort_nodes: Vec<NodeIndex>,
+    // The index of a node in the topological order (the inverse of topsort_nodes)
+    topsort_ind: UnmanagedDenseMap<NodeIndex, usize>,
+    // A temporary datastructure used during `is_convex`
+    causal: CausalVec,
+}
+
+impl<G> ConvexChecker<G>
+where
+    G: LinkView + Copy,
+{
+    /// Create a new ConvexChecker.
+    pub fn new(graph: G) -> Self {
+        let mut topsort_nodes = Vec::with_capacity(graph.node_count());
+        while topsort_nodes.len() < graph.node_count() {
```

Comment:
Can't you traverse `graph.nodes_iter()` once, get nodes with no inputs, and run toposort from there?

Reply:
True, sorry I'm getting tired. I mixed up two separate concerns: the only thing I needed to take care of is the case where there are inputs put they are not connected.